omero_certificates/certificates.py

```diff
@@ -45,81 +56,82 @@ def set_if_empty(cfgkey, default):
     return cfgdict
 
 
-def run_openssl(args):
-    command = ["openssl"] + args
-    log.info("Executing: %s", " ".join(command))
-    subprocess.run(command)
-
-
 def create_certificates(omerodir):
     cfgmap = update_config(omerodir)
     certdir = cfgmap["omero.glacier2.IceSSL.DefaultDir"]
 
     cn = cfgmap["omero.certificates.commonname"]
     owner = cfgmap["omero.certificates.owner"]
-    days = "365"
+    days = 365
     pkcs12path = os.path.join(certdir, cfgmap["omero.glacier2.IceSSL.CertFile"])
     keypath = os.path.join(certdir, cfgmap["omero.certificates.key"])
     certpath = os.path.join(certdir, cfgmap["omero.glacier2.IceSSL.CAs"])
     password = cfgmap["omero.glacier2.IceSSL.Password"]
 
-    try:
-        run_openssl(["version"])
-    except subprocess.CalledProcessError as e:
-        msg = "openssl version failed, is it installed?"
-        log.fatal("%s: %s", msg, e)
-        raise
-
     os.makedirs(certdir, exist_ok=True)
     created_files = []
 
     # Private key
     if os.path.exists(keypath):
         log.info("Using existing key: %s", keypath)
+        with open(keypath, "rb") as pem_openssl_key:
+            rsa_private_key = serialization.load_pem_private_key(
+                pem_openssl_key.read(),
+                password=None,
+            )
     else:
         log.info("Creating self-signed CA key: %s", keypath)
-        run_openssl(["genrsa", "-out", keypath, "2048"])
+        # Do what `openssl genrsa -out <keypath> <numbits>` would do
+        rsa_private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
+        with open(keypath, "wb") as pem_openssl_key:
+            pem_openssl_key.write(
+                rsa_private_key.private_bytes(
+                    Encoding.PEM,
+                    PrivateFormat.TraditionalOpenSSL,  # Essentially PKCS#1
+                    NoEncryption(),
+                )
+            )
         created_files.append(keypath)
 
     # Self-signed certificate
     log.info("Creating self-signed certificate: %s", certpath)
-    run_openssl(
-        [
-            "req",
-            "-new",
-            "-x509",
-            "-subj",
-            "{}/CN={}".format(owner, cn),
-            "-days",
-            days,
-            "-key",
-            keypath,
-            "-out",
-            certpath,
-            "-extensions",
-            "v3_ca",
-        ]
+    # Do what `openssl req -x509 ...` would do
+    utcnow = datetime.utcnow()
+    try:
+        subject = issuer = x509.Name.from_rfc4514_string("{},CN={}".format(owner, cn))
```

Comment:
Confirmed that trying to run an upgraded `omero certificates` against a server configured with `omero certificates 0.2` will lead to this error.

Since I expect in 99% of the cases that the properties will be the default ones, should the code try to auto-upgrade i.e. 1- perform some string replacement, 2- test the X509 name creation from the upgraded name and 3- update the omero property if successful. 
If this auto-upgrade fails, the error message can be sent to ask the user to manually modify its owner.

Reply:
Note this upgrade issue would also be completely addressed by point 4 of https://github.com/ome/omero-certificates/issues/26#issuecomment-1278654681